File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # create dummy to store answer
        dummy = ListNode()
        node = dummy
        
        flag = 0  # 1 if sum of two digits is greater than/equal to 10
        while l1 or l2 or flag:
            # add digit
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            val = val1 + val2 + flag
            flag = val // 10
            val %= 10  # last digit
            node.next = ListNode(val)
            
            # update pointers
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            node = node.next
        
        return dummy.next

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // create dummy to store answer
        ListNode* dummy = new ListNode();
        ListNode *node = dummy;
        
        int flag = 0;  // 1 if sum >= 10
        while (l1 != nullptr || l2 != nullptr || flag) {
            // add numbers
            int val1 = (l1 != nullptr) ? l1->val : 0;
            int val2 = (l2 != nullptr) ? l2->val : 0;
            int val = val1 + val2 + flag;
            flag = val / 10;
            val %= 10;
            node->next = new ListNode(val);
            
            // update pointers
            l1 = (l1 != nullptr) ? l1->next : nullptr;
            l2 = (l2 != nullptr) ? l2->next : nullptr;
            node = node->next;
        }
        return dummy->next;
    }
};